libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -134,5 +136,20 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(StorageLayoutSpec
 	}
 
 	solAssert(baseSlotExpressionType->isImplicitlyConvertibleTo(*TypeProvider::uint256()));
-	_storageLayoutSpecifier.annotation().baseSlot = u256(rationalType->value().numerator());
+	bigint baseSlot = u256(rationalType->value().numerator());
+	storageLayoutSpecifier->annotation().baseSlot = u256(baseSlot);
+
+	bigint size = 0;
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+		for (VariableDeclaration const* variable: contract->stateVariables())
+			if (!(variable->isConstant() || variable->immutable()))
+				size += variable->annotation().type->storageSizeUpperBound();
+
+	solAssert(size < bigint(1) << 256);
+	if (baseSlot + size >= bigint(1) << 256)
```

Comment:
Size of 2^256 is still fine. Don't we have a test with something of that size?

```suggestion
	solAssert(size <= bigint(1) << 256);
	if (baseSlot + size > bigint(1) << 256)
```

Reply:
Wait, looks like the length is actually capped at 2^256 - 1: https://github.com/ethereum/solidity/blob/5879562d17e181e53ad7adc29374dae21c863c04/libsolidity/analysis/DeclarationTypeChecker.cpp#L356-L361

Not sure why though. Is it a bug? Such a large array would be problematic for many reasons, but I don't see a good reason to arbitrarily choose a limit that's not the theoretical maximum, but one less.

Not sure if there's a use case for putting something in the very last slot but this limitation prevents it.

Reply:
Now that I think of it, the restriction against the size still does not prevent us from making the last slot reachable for smaller objects. We'll need to discuss that.

Reply:
We decided to lift the restriction, but it will be done in a separate PR: #15911.